README.md
tv.jquery
=========

TV/HbbTV development supporting environment. jQuery integrated.


 Collects basic device info, provides additional level for development (additional pseudo types for link tag media).
 opera-tv, webkit-tv, css3ui, loewe, philips, smartbox, technisat, smartbox, videoweb and to be extended.

Interface

 Invokes head link[media=loewe|smartbox|other manufactirer|css3ui|opera-tv|webkit-tv] stylefiles automatically
 according to device info accessible.

 jQuery.tv.isTv     bool

 jQuery.tv.isHbbtv  bool

 jQuery.tv.type     string   Device manufacturer/type

 jQuery.tv.id       string   identifier/serial number, 'unknown' if no data can be fetched/generated.

 jQuery.tvinfo();   TV info as single string (bult from jQuery.tv.* attributes)

Usage

 - Load jQuery
 - Load this plugin
 - Load rest scripts

 CSS3 UI:                       <link media="opera-tv css3ui" src="...css3 ui navigation..." />
 Manufacter-specific style/fix: <link media="loewe"           src="...loewe-specific CSS/fix..." />
 Browser-specific style/fix:    <link media="webkit-tv"       src="...webkit-specific CSS/fix..." />

 $(document).tvready(function(){
     ... your code here ...
 });
 $(document).bind('tvready', function(){
      ... your code here ...
  });

TODO:

   - Check if embeddeded object nodes working without being added to document, as CSS3 UI test
   - provide back log of all ids (serial, deviceid, MAC, localStorage)
   - test invoking medias before document.ready
   - document keyup handle
   -- 'numeric' fields, that has automatic mapping characters to numbers as on remote control (delayed init)
   -- fix colored buttons code insertion in text controls (yellow operates as backspace)


CHANGELOG
08-Mar-2013: Console and media plugins downgraded to use clear JS API. Left jQuery AJAX and Events integration.
             Code structure simplified.

tv.media.js
(function ($) {
    var getTvVendorByUserAgent = function (strUserAgent) {
        var patterns = {
            'loewe':      'link[media*="loewe"]',
            'smart;':     'link[media*="smartbox"]',
            'videoweb;':  'link[media*="videoweb"]',
            'technisat':  'link[media*="technisat"]',
            'philips':    'link[media*="philips"]',
            'nettv':      'link[media*="philips"]',
            'lg netcast': 'link[media*="lgelectronics"]'
        };

        for (var p in patterns) {
            if (strUserAgent.indexOf(p) >= 0) {
                return patterns[p];
            }
        }

        return undefined;
    };

    var getMediasForLoad = function (strUserAgent) {
        var arrMediasForLoad = [];

        var strMediaVendorSelector = getTvVendorByUserAgent(strUserAgent);
        if (strMediaVendorSelector != undefined) {
            arrMediasForLoad.push(strMediaVendorSelector);
        }

        if (strUserAgent.indexOf('opera') >= 0) {
            arrMediasForLoad.push('link[media*="opera-tv"]');
            return arrMediasForLoad;
        }
        if (strUserAgent.indexOf('webkit') >= 0) {
            arrMediasForLoad.push('link[media*="webkit-tv"]');
            return arrMediasForLoad;
        }

        return arrMediasForLoad;
    };

    var runPlugin = function () {
        headNode = document.getElementsByTagName('head')[0];

        var arrMediasForLoad = getMediasForLoad(navigator.userAgent.toLowerCase());
        arrMediasForLoad = arrMediasForLoad.join(', ');

        var strLogMessage = 'Pseudo medias: ' + arrMediasForLoad;
        window.console.log(strLogMessage);

        var links = headNode.querySelectorAll(arrMediasForLoad);
        for (var i = links.length - 1; i >= 0 ; --i) {
            links[i].setAttribute('media', 'screen');
        }
    };

    var headNode;
    $(document).ready(runPlugin);

})(jQuery);

tv.console.js
/**
 * @author Vladimir Reznichenko <kalessil@gmail.com>
 * @date   14.12.2012
 *
 * https://github.com/kalessil/tv.jquery
 *
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.opensource.org/licenses/GPL-2.0
 *
 * This script provides simple error console on TV devices
 */
(function ($) {
    var consoleBuffer = [];

    var consoleBufferEntry_toString = function (entry) {
        var strMessage = entry[0];
        var countAppeared = entry[1];

        var strReportAppeared = '';
        if (countAppeared > 1) {
            strReportAppeared = ' [' + countAppeared + ']';
        }

        return '&rsaquo; ' + strMessage + strReportAppeared;
    };


    var consoleBuffer_clear = function () {
        while (consoleBuffer.length > 1) {
            consoleBuffer.pop();
        }
    };
    var consoleBuffer_hasDuplicates = function (strMessage) {
        var isDuplicate;
        var entry;

        for (var position in consoleBuffer) {
            entry = consoleBuffer[position];
            isDuplicate = (entry[0] === strMessage);
            if (isDuplicate) {
                return true;
            }
        }

        return false;
    };
    var consoleBuffer_push = function (strMessage) {
        var newEntry = [strMessage, 1];
        consoleBuffer.push(newEntry);
    };
    var consoleBuffer_popDuplicateUp = function (strMessage) {
        var isDuplicate;
        var entry;

        for (var position in consoleBuffer) {
            entry = consoleBuffer[position];
            isDuplicate = (entry[0] === strMessage);
            if (isDuplicate) {
                var lastEntryPosition = consoleBuffer.length - 1;

                if (lastEntryPosition != position) {
                    var backupLastOne = consoleBuffer[lastEntryPosition];
                    consoleBuffer[lastEntryPosition] = consoleBuffer[position];
                    consoleBuffer[position] = backupLastOne;
                }

                return;
            }
        }
    };
    var consoleBuffer_count = function() {
        return consoleBuffer.length;
    };
    var consoleBuffer_mergeDuplicate = function (strMessage) {
        var isDuplicate;
        var entry;

        for (var position in consoleBuffer) {
            entry = consoleBuffer[position];
            isDuplicate = (entry[0] === strMessage);
            if (isDuplicate) {
                ++consoleBuffer[position][1];
                return;
            }
        }
    };

    var consoleBuffer_registerEntry = function (strMessage) {
        var hasDuplicates = consoleBuffer_hasDuplicates(strMessage);

        if (!hasDuplicates) {
            consoleBuffer_push(strMessage);
            return;
        }

        consoleBuffer_mergeDuplicate(strMessage);
        consoleBuffer_popDuplicateUp(strMessage);
    };

    var consoleObject_create = function () {
        var consoleObject = {
            log:function (strMessage) {
                consoleBuffer_registerEntry(strMessage);
                consoleObject.refresh();
            },
            debug:function (strMessage) {
                consoleBuffer_registerEntry('Debug: ' + strMessage);
                consoleObject.refresh();
            },
            error:function (strMessage) {
                consoleBuffer_registerEntry('<span style="color:#F00">' + strMessage + '</span>');
                consoleObject.refresh();
            },
            info:function (strMessage) {
                consoleBuffer_registerEntry('Info: ' + strMessage);
                consoleObject.refresh();
            },
            warn:function (strMessage) {
                consoleBuffer_registerEntry('<span style="color:#FF7000">' + strMessage + '</span>');
                consoleObject.refresh();
            },
            refresh:function () { },
            clear:function () {
                consoleBuffer_clear();
                consoleObject.refresh();
            }
        };

        return consoleObject;
    };

    var ajaxData_toString = function (event, jqXHR, ajaxSettings){
        var strUrl = ajaxSettings.url;
        var strStatus = jqXHR.statusText;
        var strMethod = ajaxSettings.type;

        var isHttpsOn = (window.location.protocol.toLowerCase() === 'https');
        if(isHttpsOn) {
            var ajaxProtocol = (strUrl.split(':', 1))[0];
            if(ajaxProtocol === 'http') {
                strUrl = strUrl.replace('http://', '<span style="color:#F00">http</span>://');
            }
        }

        strUrl = '<span style="color:#666;text-decoration:underline">' + strUrl + '</span>';
        strStatus = (strStatus ? (' <span style="color:#a6a6a6">' + strStatus + '</span>') : '');
        return strMethod + ' ' + strUrl + strStatus;
    };

    /* attaching listeners */
    window.addEventListener('error', function (event) {
        window.console.error(event.message);
        return true;
    });
    $(document).ajaxError(function (event, jqXHR, ajaxSettings, thrownError) {
        var strErrorMessage = ajaxData_toString(event, jqXHR, ajaxSettings);
        window.console.error(strErrorMessage + ' ' + (thrownError ? thrownError : ''));
    });
    $(document).ajaxSuccess(function (event, jqXHR, ajaxSettings) {
        var strLogMessage = ajaxData_toString(event, jqXHR, ajaxSettings);
        window.console.info(strLogMessage);
    });


    var gui_redraw = function () {
        var hasEntries = consoleBuffer_count();
        if (!hasEntries) {
            return;
        }

        gui_writeEntriesDown(consoleBuffer);
    };

    var gui_writeEntriesDown = function(arrEntries) {
        consoleDivNode.innerHTML = '';

        var strEntry;
        var consoleNode = $(consoleDivNode);
        for(var position in arrEntries) {
            strEntry = consoleBufferEntry_toString(arrEntries[position]);
            consoleNode.append(strEntry);
            consoleDivNode.appendChild(lineBreakNode.cloneNode(false));
        }

        consoleDivNode.style.display = 'block';
    };

    var consoleDivNode;
    var gui_createDiv = function () {
        var divNode = document.createElement('div');
        divNode.setAttribute('id', 'tv-console');

        divNode.style.display = 'none';
        divNode.style.position = 'absolute';
        divNode.style.bottom = 0;
        divNode.style.left = 0;
        divNode.style.width = '100%';
        divNode.style.borderTop = 'solid 1px #c6c6c6';
        divNode.style.paddingTop = '7px';
        divNode.style.paddingRight = '7px';
        divNode.style.paddingBottom = '7px';
        divNode.style.paddingLeft = '7px';
        divNode.style.color = '#a6a6a6';
        divNode.style.fontSize = '0.9em';

        document
            .getElementsByTagName('body')[0]
            .appendChild(divNode);
        consoleDivNode = divNode;
    };

    var runPlugin = function () {
        lineBreakNode = document.createElement('br');
        window.console = consoleObject_create();

        var strNativeConsoleStatus = '';
        if(typeof window.console != 'undefined') {
            strNativeConsoleStatus = ' (native one is overridden)';
        }
        consoleBuffer_registerEntry('TV console started' + strNativeConsoleStatus);

        gui_createDiv();
        window.console.refresh = gui_redraw;
        window.console.refresh();
    };

    var lineBreakNode;
    $(document).ready(runPlugin);

})(jQuery);

tv.media-min.js
/**
 * @author Vladimir Reznichenko <kalessil@gmail.com>
 * @date   12.05.2012
 *
 * https://github.com/kalessil/tv.jquery
 *
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.opensource.org/licenses/GPL-2.0
 */

(function(c){var d=function(f){var g={loewe:'link[media*="loewe"]',"smart;":'link[media*="smartbox"]',"videoweb;":'link[media*="videoweb"]',technisat:'link[media*="technisat"]',philips:'link[media*="philips"]',nettv:'link[media*="philips"]',"lg netcast":'link[media*="lgelectronics"]'};for(var h in g){if(f.indexOf(h)>=0){return g[h]}}return undefined};var e=function(f){var h=[];var g=d(f);if(g!=undefined){h.push(g)}if(f.indexOf("opera")>=0){h.push('link[media*="opera-tv"]');return h}if(f.indexOf("webkit")>=0){h.push('link[media*="webkit-tv"]');return h}return h};var b=function(){a=document.getElementsByTagName("head")[0];var h=e(navigator.userAgent.toLowerCase());h=h.join(", ");var j="Pseudo medias: "+h;window.console.log(j);var f=a.querySelectorAll(h);for(var g=f.length-1;g>=0;--g){f[g].setAttribute("media","screen")}};var a;c(document).ready(b)})(jQuery);

tv.jquery-min.js
/**
 * @author Vladimir Reznichenko <kalessil@gmail.com>
 * @date   12.05.2012
 *
 * https://github.com/kalessil/tv.jquery
 *
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.opensource.org/licenses/GPL-2.0
 */

(function(h){h.fn.extend({tvready:function(i){h(this).bind("tvready",i)}});var b=false;var d=function(i){if(typeof i!="undefined"){b=i}if(b==="wait"){return}h(document).ready(function(){h(document).trigger("tvready");window.console.log("Device: "+h.tv.type+" #"+h.tv.id)})};var f={unknown:"unknown",loewe:"Loewe",smartbox:"smartBox",videoweb:"VideoWEB",technisat:"TechniSat",philips:"Philips",lge:"LG Electronics"};var a=false;var e=false;var g=document.querySelector("head");var c=function(){var i=function(){d("wait");h(document).ready(function(){if(!document.getElementById("jquery-tv-hwdetect")){var l=document.createElement("div");l.setAttribute("id","jquery-tv-hwdetect");l.setAttribute("style","display:none; position:absolute; left:0px; top:0px; width:0px; height:0px;");var p=document.createElement("object");p.setAttribute("id","oipf-config");p.setAttribute("type","application/oipfConfiguration");l.appendChild(p);var n=document.createElement("object");n.setAttribute("id","lgNetCastDevice");p.setAttribute("type","application/x-netcast-info");l.appendChild(n);document.querySelector("body").appendChild(l);if(h.browser.opera){a=true}if(!a){var k=document.createElement("input");k.setAttribute("id","css3ui-nav-test");k.setAttribute("type","text");k.setAttribute("style","nav-up:#css3ui-nav-test;nav-down:#css3ui-nav-test;");a=(typeof k.style.navDown!="undefined"&&k.style.navDown=="#css3ui-nav-test")}h.tv.hasSpatialNav=a;var o=navigator.cookieEnabled;if(!o&&typeof navigator.cookieEnabled=="undefined"){document.cookie="testcookie";o=(document.cookie.indexOf("testcookie")!=-1)}e=o}var s=function(y){y=y.toLowerCase().split(":");var v="";for(var x=0,w=0;x<y.length;++x){w=parseInt(y[x],16);v+=(w>99?w:(w>9?"0"+w:"00"+w))}return v};var r=function(){if(!e){return f.unknown}var z="__";var w=function(){if(typeof localStorage!="undefined"&&localStorage.getItem("deviceId")!=null){return localStorage.getItem("deviceId")}var E=document.cookie.split("; ");for(var D=0,F;(F=E[D]&&E[D].split("="));D++){if(F.shift()==="deviceId"){if(typeof localStorage!="undefined"){localStorage.setItem("deviceId",F.join("="))}return F.join("=")}}return null};var B=w();if(B!=null){return z+s(B)}var v=[],A="0123456789abcdef";for(var y=0,x="";y<6;y++){x=A.charAt(Math.floor(Math.random()*A.length));x+=A.charAt(Math.floor(Math.random()*A.length));v.push(x)}v=v.join(":");var C=function(E){if(typeof localStorage!="undefined"){localStorage.setItem("deviceId",E);return}var D=new Date();D.setDate(D.getDate()+365*25);document.cookie=["deviceId","=",E,"; expires="+D.toUTCString(),"","",""].join("")};C(v);return z+s(v)};var t=function(v){if(v==f.unknown){return v}switch(h.tv.type){case f.loewe:v="LO"+v;break;case f.smartbox:v="SB"+v;break;case f.videoweb:v="VW"+v;break;case f.technisat:v="TS"+v;break;case f.philips:v="PH"+v;break;case f.lge:v="LG"+v;break;default:v="XX"+v;break}return v};var q=document.getElementById("oipf-config");if(typeof q.localSystem!="undefined"){var m=q.localSystem;if(typeof m.serialNumber!="undefined"){h.tv.id=t(m.serialNumber);return d(true)}if(typeof m.deviceID!="undefined"){h.tv.id=t(m.deviceID);return d(true)}if(typeof m.networkInterfaces!="undefined"){h.tv.id=t(s(m.networkInterfaces.item(0).macAddress));return d(true)}}if(typeof q.configuration!="undefined"){}if(typeof NetRangeDevice=="object"){h.tv.id=t(s(NetRangeDevice.getMACID()));return d(true)}var u=document.getElementById("lgNetCastDevice");if(typeof u.manufacturer!="undefined"){if(typeof n.serialNumber!="undefine"){h.tv.id=t(n.serialNumber);return d(true)}if(typeof n.net_macAddress!="undefine"){h.tv.id=t(s(n.net_macAddress));return d(true)}}h.tv.id=t(r());return d(true)})};var j=function(){var l={Loewe:f.loewe,"smart;":f.smartbox,"videoweb;":f.videoweb,TechniSat:f.technisat,Philips:f.philips,NetTV:f.philips,"LG NetCast":f.lge};var k=navigator.userAgent.toLowerCase();for(var m in l){if(k.indexOf(m.toLowerCase())>=0){return h.tv.type=l[m]}}return f.unknown};j();i();h.tv.isTv=(h.tv.isHbbtv||h.tv.type!=f.unknown||navigator.userAgent.indexOf("CE-HTML")>=0)};h.extend({tv:{isTv:false,isHbbtv:(navigator.userAgent.indexOf("HbbTV")>=0),type:f.unknown,id:f.unknown,hasSpatialNav:false}});c();h.extend({tvinfo:function(){if(b!=true){return"TV plugin is starting (concern using $(document).tvready(callback));"}var i=[];if(h.tv.isTv){i.push("TV/TV box")}if(h.tv.isHbbtv){i.push("HbbTV")}var j=document.getElementById("oipf-config");if(typeof j.localSystem!="undefined"){i.push("OIPF config (HW info)")}if(typeof j.configuration!="undefined"){i.push("OIPF config (basic)")}if(a){i.push("CSS3 UI nav-*")}if(e){i.push("Cookies")}if(typeof localStorage!="undefined"){i.push("localStorage")}if(typeof sessionStorage!="undefined"){i.push("sessionStorage")}if(h.browser.webkit){i.push("WebKit")}if(h.browser.opera){i.push("Opera")}i=(i.length>0?(" ("+i.join(", ")+")"):"");return h.tv.type+i+" #"+h.tv.id}});d()})(jQuery);

tv.jquery.js
/**
 * @author Vladimir Reznichenko <kalessil@gmail.com>
 * @date   12.05.2012
 *
 * https://github.com/kalessil/tv.jquery
 *
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.opensource.org/licenses/GPL-2.0
 */

/* Run jQuery plugin now */
(function($){
    /* immediate public interface */
    $.fn.extend({
        /**
         * Adds support for tvready event bind proxy function
         * @param callback function
         */
        tvready: function(callback) {
            $(this).bind('tvready', callback);
        }
    });
    /* immediate public interface */

    var __tv_ready_status = false;

    /**
     * Triggers TV ready event
     */
    var __ready = function(newState){
        if(typeof newState != 'undefined') {
            __tv_ready_status = newState;
        }

        if(__tv_ready_status === 'wait') {
            return;
        }

        $(document).ready(function(){
            $(document).trigger('tvready');
            window.console.log('Device: ' + $.tv.type + ' #' + $.tv.id);
        });
    };

    /* constants */
    var deviceType = {
        unknown:   'unknown',
        loewe:     'Loewe',
        smartbox:  'smartBox',
        videoweb:  'VideoWEB',
        technisat: 'TechniSat',
        philips:   'Philips',
        lge:       'LG Electronics'
    };

    var __css3uiNavSupported = false;
    var __cookiesAvailable = false;


    /* variable to be used later in several places, reducing searches */
    var __headNode = document.querySelector('head');

    var __grabInfo = function() {
        var __grabHardwareId = function(){
            /*
             * embeddable object based detection, it non immediate due to
             * domManipulations possible only on document ready
             */
            __ready('wait');

            $(document).ready(function(){

                if(!document.getElementById('jquery-tv-hwdetect'))
                {
                    /**
                     * Inject HTML for futher detection, used direct
                     * domElement creation, instead of pure jQuery, which slow.
                     * Since we are on TV, it's critical.
                     */
                    var container = document.createElement('div');
                    container.setAttribute('id', 'jquery-tv-hwdetect');
                    container.setAttribute('style', 'display:none; position:absolute; left:0px; top:0px; width:0px; height:0px;');

                    /* oipf configuration standart injection */
                    var oipfConfig = document.createElement('object');
                    oipfConfig.setAttribute('id', 'oipf-config'); oipfConfig.setAttribute('type', 'application/oipfConfiguration');
                    container.appendChild(oipfConfig);

                    var lgNetCastDevice = document.createElement('object');
                    lgNetCastDevice.setAttribute('id', 'lgNetCastDevice'); oipfConfig.setAttribute('type', 'application/x-netcast-info');
                    container.appendChild(lgNetCastDevice);

                    /* Modify document now with assembled injection */
                    document.querySelector('body').appendChild(container);

                    /*
                     * CSS3 UI spatial navigation test
                     * Check if data from inline existing is Node style attributes
                    **/
                    if($.browser.opera) {
                        __css3uiNavSupported = true;
                    }
                    if(!__css3uiNavSupported) {
                        var css4uiNav = document.createElement('input');
                        css4uiNav.setAttribute('id', 'css3ui-nav-test');
                        css4uiNav.setAttribute('type', 'text');
                        css4uiNav.setAttribute('style', 'nav-up:#css3ui-nav-test;nav-down:#css3ui-nav-test;');
                        /* Do test */
                        __css3uiNavSupported = (typeof css4uiNav.style.navDown != 'undefined' && css4uiNav.style.navDown == '#css3ui-nav-test');
                    }
                    $.tv.hasSpatialNav = __css3uiNavSupported;

                    /* Test cookie enabled */
                     var cookieEnabled = navigator.cookieEnabled;
                     if (!cookieEnabled && typeof navigator.cookieEnabled == 'undefined')
                     { 
                          document.cookie = 'testcookie';
                          cookieEnabled = (document.cookie.indexOf('testcookie') != -1);
                     }
                     __cookiesAvailable = cookieEnabled;
                     /* Test cookie enabled */
                }
                
                
                /* I don't like using raw MAC, generate some kind of serial from it */
                var __toSerial = function(mac) {
                    mac = mac.toLowerCase().split(':');

                    var result = '';
                    for(var i = 0, tmp = 0; i < mac.length; ++i)
                    {
                        /* each byte to decimal basis in XXX format*/
                        tmp = parseInt(mac[i], 16);
                        result += (tmp > 99 ? tmp : (tmp > 9 ? '0' + tmp : '00' + tmp));
                    }
                    return result;
                };

                /* Generate pseudo MAC address */
                var __serialNumberSelfAssign = function() {
                    if(!__cookiesAvailable) {
                        return deviceType.unknown;
                    }
                    
                    var prefix = '__';
                    
                    /* check cookies first */
                    var __readStored = function() {
                        /* if has local storage support */
                        if(typeof localStorage != 'undefined' && localStorage.getItem('deviceId') != null) {
                            return localStorage.getItem('deviceId');
                        }
                        
                        /* use cookies */
                        var cookies = document.cookie.split('; ');
                        for (var i = 0, parts; (parts = cookies[i] && cookies[i].split('=')); i++) {
                            if (parts.shift() === 'deviceId') {
                                /* Provide migration cookies -> local storage when soft upgraded */
                                if(typeof localStorage != 'undefined') {
                                    localStorage.setItem('deviceId', parts.join('='));
                                }

                                return parts.join('=');
                            }
                        }
                        return null;
                    };
                    var cooked = __readStored();
                    if(cooked != null) {
                        return prefix + __toSerial(cooked);
                    }

                    /* generate new MAC */
                    var result = [], alphabet = '0123456789abcdef';
                    for (var i = 0, part = ''; i < 6; i++)
                    {
                       /* generate single byte in hex basis */
                       part  = alphabet.charAt(Math.floor(Math.random() * alphabet.length));
                       part += alphabet.charAt(Math.floor(Math.random() * alphabet.length));
                       result.push(part);
                    }
                    result = result.join(':');
                    /* generate new MAC */

                    /* store it */
                    var __writeStored = function(mac) {
                        /* if has local storage support */
                        if(typeof localStorage != 'undefined') {
                            localStorage.setItem('deviceId', mac);
                            return;
                        }
                        
                        /* use cookies */
                        var expire = new Date(); expire.setDate(expire.getDate() + 365 * 25);
                        /* 25 years is enought, earlier soft update can drop cookies */
                        document.cookie = [
                               'deviceId', '=', mac, /* data */
                               '; expires=' + expire.toUTCString(),
                               '', /* path */
                               '', /* domain */
                               ''  /* secure */
                           ].join('');
                    };
                    __writeStored(result);

                    return prefix + __toSerial(result);
                };

                /* Will add prefix disclosuring manufacturer to given serial */
                var __prefixSerial = function(serial){
                    if(serial == deviceType.unknown) {
                        return serial;
                    }

                    switch($.tv.type)
                    {
                       case deviceType.loewe:
                           serial = 'LO' + serial;
                           break;
                       case deviceType.smartbox:
                           serial = 'SB' + serial;
                           break;
                       case deviceType.videoweb:
                           serial = 'VW' + serial;
                           break;
                       case deviceType.technisat:
                           serial = 'TS' + serial;
                           break;
                       case deviceType.philips:
                           serial = 'PH' + serial;
                           break;
                       case deviceType.lge:
                           serial = 'LG' + serial;
                           break;
                       
                       default:
                           serial = 'XX' + serial;
                           break;
                    }

                    return serial;
                };
                
                /* @see oipf.tv, primary detection */
                var oipfObject = document.getElementById('oipf-config');
                if (typeof oipfObject.localSystem != 'undefined') 
                {
                    var hwInfo = oipfObject.localSystem;

                    /* TODO: how handle extention of supporting festures on soft upgrade? */

                    /* Fine, we have serial from manufacturer */
                    if(typeof hwInfo.serialNumber != 'undefined') {
                        $.tv.id = __prefixSerial(hwInfo.serialNumber);
                        return __ready(true);
                    } 

                    /* Device ID is acceptable as well */
                    if(typeof hwInfo.deviceID != 'undefined') {
                        $.tv.id = __prefixSerial(hwInfo.deviceID);
                        return __ready(true);
                    } 

                    /* Ok, then identify by first Network Interface (MAC address) */
                    if(typeof hwInfo.networkInterfaces != 'undefined') {
                       $.tv.id = __prefixSerial(__toSerial(hwInfo.networkInterfaces.item(0).macAddress));
                       return __ready(true);
                    }
                }
                
                if(typeof oipfObject.configuration != 'undefined') {
                    /* 
                     * Only basics are implemented on current device, no LocalSystem.
                     *
                     * TODO: discuss if langs to be extracted to $.tv container
                     **/
                }

                /* Non-standard capabilities */
                /* Loewe */
                if(typeof NetRangeDevice == 'object') {
                    /* Loewe */
                    $.tv.id = __prefixSerial(__toSerial(NetRangeDevice.getMACID()));
                    return __ready(true);
                }
                /* LG */
                var lgNetCastDeviceObject = document.getElementById('lgNetCastDevice');
                if(typeof lgNetCastDeviceObject.manufacturer != 'undefined') {
                    if(typeof lgNetCastDevice.serialNumber != 'undefine') {
                        $.tv.id = __prefixSerial(lgNetCastDevice.serialNumber);
                        return __ready(true);
                    }

                    if(typeof lgNetCastDevice.net_macAddress != 'undefine') {
                        $.tv.id = __prefixSerial(__toSerial(lgNetCastDevice.net_macAddress));
                        return __ready(true);
                    }
                }
                /* Non-standard capabilities */


                /* failed do detect anything, fallback to fake MAC id */
                $.tv.id = __prefixSerial(__serialNumberSelfAssign());
                return __ready(true);
            });
        };

        /**
         * Check UA to define who is manufacturer/what is device
         * @return string
         */
        var __grabHardwareType = function()
        {
            /* loop on object to match strings */
            var patterns = {
                'Loewe':      deviceType.loewe,
                'smart;':     deviceType.smartbox,
                'videoweb;':  deviceType.videoweb,
                'TechniSat':  deviceType.technisat,
                'Philips':    deviceType.philips,   'NetTV': deviceType.philips,
                'LG NetCast': deviceType.lge
            };
            
            var ua = navigator.userAgent.toLowerCase();
            for(var p in patterns) {
                if (ua.indexOf(p.toLowerCase()) >= 0) {
                    return $.tv.type = patterns[p];
                }
            }

            /* default result is unknown */
            return deviceType.unknown;
        };

        __grabHardwareType();
        __grabHardwareId();

        $.tv.isTv = ($.tv.isHbbtv || $.tv.type != deviceType.unknown || navigator.userAgent.indexOf('CE-HTML') >= 0);
    };


    /* populate data storage */
    $.extend({
        tv: {
            'isTv':    false,
            'isHbbtv': (navigator.userAgent.indexOf('HbbTV') >= 0),
            'type':    deviceType.unknown,
            'id':      deviceType.unknown,
            'hasSpatialNav': false
        }
    });
    __grabInfo();
    /* populate data storage */


    /* public interface */
    $.extend({
        /**
         * Builds human readable device information.
         *
         * @return string
         */
        tvinfo: function() {
            if (__tv_ready_status != true) {
                return 'TV plugin is starting (concern using $(document).tvready(callback));';
            }

            /* Features report */
            var features = [];
            if($.tv.isTv) {                                         features.push('TV/TV box'); }
            if($.tv.isHbbtv) { features.push('HbbTV'); }

            var localSystem = document.getElementById('oipf-config');
            if (typeof localSystem.localSystem != 'undefined') {    features.push('OIPF config (HW info)'); }
            if (typeof localSystem.configuration != 'undefined') {  features.push('OIPF config (basic)'); }

            if(__css3uiNavSupported) {                              features.push('CSS3 UI nav-*'); }
            if(__cookiesAvailable) {                                features.push('Cookies'); }
            if(typeof localStorage != 'undefined') {                features.push('localStorage'); }
            if(typeof sessionStorage != 'undefined') {              features.push('sessionStorage'); }
            if($.browser.webkit) {                                  features.push('WebKit'); }
            if($.browser.opera) {                                   features.push('Opera'); }

            features = (features.length > 0 ? (' (' + features.join(', ') + ')') : '');
            /* Features report */


            return $.tv.type + features + ' #' + $.tv.id;
        }
    });
    /* public interface */


    /* Finalize plugin load */
    __ready();
    /* Finalize plugin load */

})(jQuery);